packages/aws-rfdk/lib/core/lambdas/nodejs/version-provider/handler.ts

```diff
@@ -0,0 +1,357 @@
+
+        if (versionedUris) {
+          installers.set(p, versionedUris);
+        }
+      });
+    }
+
+    return installers;
+  }
+
+  /**
+   * @inheritdoc
+   */
+  /* istanbul ignore next */ // @ts-ignore
+  public async doDelete(physicalId: string, resourceProperties: IVersionProviderProperties): Promise<void> {
+    // Nothing to do -- we don't modify anything.
+    return;
+  }
+
+  private implementsIVersionProviderProperties(value: any): boolean {
+    if (!value || typeof(value) !== 'object') { return false; }
+
+    // TODO: Ignore case
+    if (!value.product || !Object.values(Product).includes(value.product)) {
+      return false;
+    }
+
+    if (value.versionString) {
+      if (null === this.parseVersionString(value.versionString))  { return false; }
+    }
+
+    if (value.platform) {
+      // TODO: Ignore case
+      if (!Object.values(Platform).includes(value.platform))  { return false; }
+    }
+
+    return true;
+  }
+
+  /* istanbul ignore next */ // @ts-ignore
+  private async downloadInstallerIndex() {
+    const productionInfoURL = 'https://downloads.thinkboxsoftware.com/version_info.json';
+
+    // eslint-disable-next-line @typescript-eslint/no-require-imports
+    const parsedUrl = require('url').parse(productionInfoURL);
+
+    const options = {
+      host: parsedUrl.hostname,
+      path: parsedUrl.path,
+    };
+
+    return new Promise((resolve, reject) => {
+      // eslint-disable-next-line @typescript-eslint/no-require-imports
+      require('https').get(options, (res: IncomingMessage) => {
+        let json = '';
+
+        res.on('data', (chunk: any) => {
+          // keep appending the response chunks until we get 'end' event.
+          json += chunk;
+        });
+
+        res.on('end', () => {
+          // complete response is available here:
+          if (res.statusCode === 200) {
+            try {
+              // convert the response to a json object and return.
+              const data = JSON.parse(json);
+              resolve(data);
+            } catch (e) {
+              reject(e);
+            }
+          } else {
+            reject(new Error(`Expected status code 200, but got ${res.statusCode}`));
+          }
+        });
+      }).on('error', (err: Error) => {
+        reject(err);
+      });
+    });
+  }
+
+  /**
+   * This method reads index file.
+   */
+  private readInstallersIndex(): any {
+    if (!this.indexFilePath) {
+      throw new Error('File path should be defined.');
+    }
+    if (!fs.existsSync(this.indexFilePath)) {
+      throw new Error(`File ${this.indexFilePath} was not found`);
+    }
+    const data = fs.readFileSync(this.indexFilePath, 'utf8');
+
+    // convert the response to a json object and return.
+    const json = JSON.parse(data);
+    return json;
+  }
+
+  private parseVersionString(versionString: string): RegExpExecArray | null {
+    const VALID_VERSION_REGEX = /^(0|[1-9]\d*)(?:\.(0|[1-9]\d*))?(?:\.(0|[1-9]\d*))?(?:\.(0|[1-9]\d*))?$/;
```

Comment:
Since this REGEX is constant could we make this a class level constant.

Also could we use /^(0|[1-9]\d*)(?:\.(0|[1-9]\d*)){0,3}$/ instead. the {0,3} means 0-3 copies of the preceeding so we can use this instead of repeating the same section 3 times with the ? quantifier.

Reply:
Your expression work great for validation but it return only first and last part as result of `exec`